[consensus] Add DagBuilder & DagParser V1

arun-koshy opened the pull request:
## Description 

Add `DagBuilder` API and Dag DSL that can be parsed via `DagParser` module.

### DagBuilder API
```
let dag_builder = DagBuilder::new(num_authorities); 
dag_builder.layer(1).build(); // Round 1 is fully connected with parents by default.
dag_builder.layers(2..10).build(); // Rounds 2 ~ 10 are fully connected with parents by default.
dag_builder.layers(11).min_parent_links().build(); // Round 11 is minimally and randomly connected with parents, without weak links.
dag_builder.layers(12).no_leader_block(0).build(); // Round 12 misses leader block. Other blocks are fully connected with parents.
dag_builder.layers(13).no_leader_link(12, 0).build(); // Round 13 misses votes for leader block. Other blocks are fully connected with parents.
dag_builder.layers(14).authorities(vec![3,5]).skip_block().build(); // Round 14 authorities 3 and 5 will not propose any block.
dag_builder.layers(15).authorities(vec![3,5]).skip_ancestor_links(vec![1,2]).build(); // Round 15 authorities 3 and 5 will not link to ancestors 1 and 2
dag_builder.layers(16).authorities(vec![3,5]).equivocate(3).build(); // Round 16 authorities 3 and 5 will produce 3 equivocating blocks.
```
- calling `.build()` will write blocks to dag state.
- todo: finish implementation for `random_weak_links()`
- `dag_builder.blocks `has all the blocks to be retrieved for additional testing
- `dag_builder.print()` will pretty print the constructed dag, i.e.
```
DAG {
Round 1 : 
    Block A1(otYgq5ty3gLz5XNmp+gkj6lRsr1MXHPH8eUrB5TRqSw=)(1001ms;[A0(XQ0/ISCOxrPDK4Xl1TW4BHE797ZYVZoQBYycTZ/Sx5o=), B0(O0Mjg1TlElvBC9yfxJDZeSXnGZopE6Qn8Cv6xTRVRx4=), C0(W/YWKrKixuuhVd+ucuyGHegk9gSkObZiRxdrYq1ijm4=), D0(pwrloftpTl5TN3/5UyvH/Ku0oVU1PxpkNzg73Uz+rIA=)];0;v)
    Block B1(FnRBfqZV0pNQO2rL1XbLnijISvJPWSaFX3o35CY11+c=)(1002ms;[A0(XQ0/ISCOxrPDK4Xl1TW4BHE797ZYVZoQBYycTZ/Sx5o=), B0(O0Mjg1TlElvBC9yfxJDZeSXnGZopE6Qn8Cv6xTRVRx4=), C0(W/YWKrKixuuhVd+ucuyGHegk9gSkObZiRxdrYq1ijm4=), D0(pwrloftpTl5TN3/5UyvH/Ku0oVU1PxpkNzg73Uz+rIA=)];0;v)
    Block C1(PCIpbD4F5jgZNj4SEHpiQNDkihl8mhkhHQfhBPW1LsA=)(1003ms;[A0(XQ0/ISCOxrPDK4Xl1TW4BHE797ZYVZoQBYycTZ/Sx5o=), B0(O0Mjg1TlElvBC9yfxJDZeSXnGZopE6Qn8Cv6xTRVRx4=), C0(W/YWKrKixuuhVd+ucuyGHegk9gSkObZiRxdrYq1ijm4=), D0(pwrloftpTl5TN3/5UyvH/Ku0oVU1PxpkNzg73Uz+rIA=)];0;v)
    Block D1(FXa/64rtmX4fHea21XL1rm28xFib3IDgSw6zHpNvWNY=)(1004ms;[A0(XQ0/ISCOxrPDK4Xl1TW4BHE797ZYVZoQBYycTZ/Sx5o=), B0(O0Mjg1TlElvBC9yfxJDZeSXnGZopE6Qn8Cv6xTRVRx4=), C0(W/YWKrKixuuhVd+ucuyGHegk9gSkObZiRxdrYq1ijm4=), D0(pwrloftpTl5TN3/5UyvH/Ku0oVU1PxpkNzg73Uz+rIA=)];0;v)
Round 2 : 
    Block A2(zrcEl6WAMUiT1a463jhovInUBpYSX1r9ZYsdyKe1Mrs=)(2002ms;[A1(otYgq5ty3gLz5XNmp+gkj6lRsr1MXHPH8eUrB5TRqSw=), B1(FnRBfqZV0pNQO2rL1XbLnijISvJPWSaFX3o35CY11+c=), C1(PCIpbD4F5jgZNj4SEHpiQNDkihl8mhkhHQfhBPW1LsA=), D1(FXa/64rtmX4fHea21XL1rm28xFib3IDgSw6zHpNvWNY=)];0;v)
    Block B2(Gn1w4RCXN+o5X5hsW+Z85/ZUJmD+LdGI4/vkjSCdgQI=)(2003ms;[A1(otYgq5ty3gLz5XNmp+gkj6lRsr1MXHPH8eUrB5TRqSw=), B1(FnRBfqZV0pNQO2rL1XbLnijISvJPWSaFX3o35CY11+c=), C1(PCIpbD4F5jgZNj4SEHpiQNDkihl8mhkhHQfhBPW1LsA=), D1(FXa/64rtmX4fHea21XL1rm28xFib3IDgSw6zHpNvWNY=)];0;v)
    Block C2(zXTiZ9tvD1tyZQBzEwiELizzOH4UoaDlYeHnufN6d6w=)(2004ms;[A1(otYgq5ty3gLz5XNmp+gkj6lRsr1MXHPH8eUrB5TRqSw=), B1(FnRBfqZV0pNQO2rL1XbLnijISvJPWSaFX3o35CY11+c=), C1(PCIpbD4F5jgZNj4SEHpiQNDkihl8mhkhHQfhBPW1LsA=), D1(FXa/64rtmX4fHea21XL1rm28xFib3IDgSw6zHpNvWNY=)];0;v)
    Block D2(WZV+pvnmcOQBY87nPAQZAXUGo1vs2JnYHCVNSvaOWNE=)(2005ms;[A1(otYgq5ty3gLz5XNmp+gkj6lRsr1MXHPH8eUrB5TRqSw=), B1(FnRBfqZV0pNQO2rL1XbLnijISvJPWSaFX3o35CY11+c=), C1(PCIpbD4F5jgZNj4SEHpiQNDkihl8mhkhHQfhBPW1LsA=), D1(FXa/64rtmX4fHea21XL1rm28xFib3IDgSw6zHpNvWNY=)];0;v)
Round 3 : 
    Block A3(OhYZrKR85wAXtUQwZX+Yh8IWGot6QOPdrBwIaL36+cU=)(3003ms;[A2(zrcEl6WAMUiT1a463jhovInUBpYSX1r9ZYsdyKe1Mrs=), B2(Gn1w4RCXN+o5X5hsW+Z85/ZUJmD+LdGI4/vkjSCdgQI=), C2(zXTiZ9tvD1tyZQBzEwiELizzOH4UoaDlYeHnufN6d6w=), D2(WZV+pvnmcOQBY87nPAQZAXUGo1vs2JnYHCVNSvaOWNE=)];0;v)
    Block B3(3WY5JizOmQm6yrOvwgfMFJOov5bt6VnjAPVpe3Lb2fc=)(3004ms;[A2(zrcEl6WAMUiT1a463jhovInUBpYSX1r9ZYsdyKe1Mrs=), B2(Gn1w4RCXN+o5X5hsW+Z85/ZUJmD+LdGI4/vkjSCdgQI=), C2(zXTiZ9tvD1tyZQBzEwiELizzOH4UoaDlYeHnufN6d6w=), D2(WZV+pvnmcOQBY87nPAQZAXUGo1vs2JnYHCVNSvaOWNE=)];0;v)
    Block C3(F110IA17XMGfN6zlfluVpd8fHlzlk8IfgyJFJog7R7Y=)(3005ms;[A2(zrcEl6WAMUiT1a463jhovInUBpYSX1r9ZYsdyKe1Mrs=), B2(Gn1w4RCXN+o5X5hsW+Z85/ZUJmD+LdGI4/vkjSCdgQI=), C2(zXTiZ9tvD1tyZQBzEwiELizzOH4UoaDlYeHnufN6d6w=), D2(WZV+pvnmcOQBY87nPAQZAXUGo1vs2JnYHCVNSvaOWNE=)];0;v)
    Block D3(r9Z0F6YCEQP6FIuCmYgiE/I60HAPi1qPN0xkzci3OH4=)(3006ms;[A2(zrcEl6WAMUiT1a463jhovInUBpYSX1r9ZYsdyKe1Mrs=), B2(Gn1w4RCXN+o5X5hsW+Z85/ZUJmD+LdGI4/vkjSCdgQI=), C2(zXTiZ9tvD1tyZQBzEwiELizzOH4UoaDlYeHnufN6d6w=), D2(WZV+pvnmcOQBY87nPAQZAXUGo1vs2JnYHCVNSvaOWNE=)];0;v)
Round 4 : 
    Block A4(0btWl7oo1vjr+9XlQRqCbVJzcFQwqym0MOxeD7dJ0Lo=)(4004ms;[A3(OhYZrKR85wAXtUQwZX+Yh8IWGot6QOPdrBwIaL36+cU=), B3(3WY5JizOmQm6yrOvwgfMFJOov5bt6VnjAPVpe3Lb2fc=), C3(F110IA17XMGfN6zlfluVpd8fHlzlk8IfgyJFJog7R7Y=), D3(r9Z0F6YCEQP6FIuCmYgiE/I60HAPi1qPN0xkzci3OH4=)];0;v)
    Block B4(URFOcNxMBns0CBox4WHrMTfC9Fycn7j06mPc67bRuQg=)(4005ms;[A3(OhYZrKR85wAXtUQwZX+Yh8IWGot6QOPdrBwIaL36+cU=), B3(3WY5JizOmQm6yrOvwgfMFJOov5bt6VnjAPVpe3Lb2fc=), C3(F110IA17XMGfN6zlfluVpd8fHlzlk8IfgyJFJog7R7Y=), D3(r9Z0F6YCEQP6FIuCmYgiE/I60HAPi1qPN0xkzci3OH4=)];0;v)
    Block C4(nqsRar0k6OQxmB3Gmegq34M60eInZcXdjxmKwBb4ZQ0=)(4006ms;[A3(OhYZrKR85wAXtUQwZX+Yh8IWGot6QOPdrBwIaL36+cU=), B3(3WY5JizOmQm6yrOvwgfMFJOov5bt6VnjAPVpe3Lb2fc=), C3(F110IA17XMGfN6zlfluVpd8fHlzlk8IfgyJFJog7R7Y=), D3(r9Z0F6YCEQP6FIuCmYgiE/I60HAPi1qPN0xkzci3OH4=)];0;v)
    Block D4(cLn1H9E0NaiiFRK5Am8Bm8b7NFFcdaKQbdaO9X2P3bg=)(4007ms;[A3(OhYZrKR85wAXtUQwZX+Yh8IWGot6QOPdrBwIaL36+cU=), B3(3WY5JizOmQm6yrOvwgfMFJOov5bt6VnjAPVpe3Lb2fc=), C3(F110IA17XMGfN6zlfluVpd8fHlzlk8IfgyJFJog7R7Y=), D3(r9Z0F6YCEQP6FIuCmYgiE/I60HAPi1qPN0xkzci3OH4=)];0;v)
Round 5 : 
    Block A5(2TyDCA7zQ6M6LFgHIVMPodOieoNbdwGf5CuwDEWyIxg=)(5005ms;[A4(0btWl7oo1vjr+9XlQRqCbVJzcFQwqym0MOxeD7dJ0Lo=), B4(URFOcNxMBns0CBox4WHrMTfC9Fycn7j06mPc67bRuQg=), C4(nqsRar0k6OQxmB3Gmegq34M60eInZcXdjxmKwBb4ZQ0=), D4(cLn1H9E0NaiiFRK5Am8Bm8b7NFFcdaKQbdaO9X2P3bg=)];0;v)
    Block B5(yqbdpP3wSZXR0VwQS2U9IQwHWD+RpbCRtPbNx/Csnvc=)(5006ms;[A4(0btWl7oo1vjr+9XlQRqCbVJzcFQwqym0MOxeD7dJ0Lo=), B4(URFOcNxMBns0CBox4WHrMTfC9Fycn7j06mPc67bRuQg=), C4(nqsRar0k6OQxmB3Gmegq34M60eInZcXdjxmKwBb4ZQ0=), D4(cLn1H9E0NaiiFRK5Am8Bm8b7NFFcdaKQbdaO9X2P3bg=)];0;v)
    Block C5(dUuoHsDUpEBu3kwUcrJ6xN210UiFxYanLqnLeg5suyk=)(5007ms;[A4(0btWl7oo1vjr+9XlQRqCbVJzcFQwqym0MOxeD7dJ0Lo=), B4(URFOcNxMBns0CBox4WHrMTfC9Fycn7j06mPc67bRuQg=), C4(nqsRar0k6OQxmB3Gmegq34M60eInZcXdjxmKwBb4ZQ0=), D4(cLn1H9E0NaiiFRK5Am8Bm8b7NFFcdaKQbdaO9X2P3bg=)];0;v)
    Block D5(TKqYtHz/KVYQOII4Iggr+JB4DU/0/U9GcEW+j9v/aL0=)(5008ms;[A4(0btWl7oo1vjr+9XlQRqCbVJzcFQwqym0MOxeD7dJ0Lo=), B4(URFOcNxMBns0CBox4WHrMTfC9Fycn7j06mPc67bRuQg=), C4(nqsRar0k6OQxmB3Gmegq34M60eInZcXdjxmKwBb4ZQ0=), D4(cLn1H9E0NaiiFRK5Am8Bm8b7NFFcdaKQbdaO9X2P3bg=)];0;v)
Round 6 : 
    Block A6(luf/p0Q4l8TTqfmZdi5+ioJOR8d1BuOrPGXqxs1Vh4E=)(6006ms;[A5(2TyDCA7zQ6M6LFgHIVMPodOieoNbdwGf5CuwDEWyIxg=), B5(yqbdpP3wSZXR0VwQS2U9IQwHWD+RpbCRtPbNx/Csnvc=), C5(dUuoHsDUpEBu3kwUcrJ6xN210UiFxYanLqnLeg5suyk=), D5(TKqYtHz/KVYQOII4Iggr+JB4DU/0/U9GcEW+j9v/aL0=)];0;v)
    Block B6(6LaW2zW6P4y9FOqsdh0Vx3vVrjFzbcu69iC1OxfDmMY=)(6007ms;[A5(2TyDCA7zQ6M6LFgHIVMPodOieoNbdwGf5CuwDEWyIxg=), B5(yqbdpP3wSZXR0VwQS2U9IQwHWD+RpbCRtPbNx/Csnvc=), C5(dUuoHsDUpEBu3kwUcrJ6xN210UiFxYanLqnLeg5suyk=), D5(TKqYtHz/KVYQOII4Iggr+JB4DU/0/U9GcEW+j9v/aL0=)];0;v)
    Block C6(v+SNVAiEe8m9X0W47l9tP8Zg4jUg2+RLqTSWdguLMFo=)(6008ms;[A5(2TyDCA7zQ6M6LFgHIVMPodOieoNbdwGf5CuwDEWyIxg=), B5(yqbdpP3wSZXR0VwQS2U9IQwHWD+RpbCRtPbNx/Csnvc=), C5(dUuoHsDUpEBu3kwUcrJ6xN210UiFxYanLqnLeg5suyk=), D5(TKqYtHz/KVYQOII4Iggr+JB4DU/0/U9GcEW+j9v/aL0=)];0;v)
    Block D6(T9Wk1U9aj+X+NNH5mBBsHumTHTBfN85mvjy+HQKCFPs=)(6009ms;[A5(2TyDCA7zQ6M6LFgHIVMPodOieoNbdwGf5CuwDEWyIxg=), B5(yqbdpP3wSZXR0VwQS2U9IQwHWD+RpbCRtPbNx/Csnvc=), C5(dUuoHsDUpEBu3kwUcrJ6xN210UiFxYanLqnLeg5suyk=), D5(TKqYtHz/KVYQOII4Iggr+JB4DU/0/U9GcEW+j9v/aL0=)];0;v)
Round 7 : 
    Block A7(o/lGD0z5GLCGMLo8dpWWYCMe8caXFw6fCuCU/s4bH7E=)(7007ms;[A6(luf/p0Q4l8TTqfmZdi5+ioJOR8d1BuOrPGXqxs1Vh4E=), B6(6LaW2zW6P4y9FOqsdh0Vx3vVrjFzbcu69iC1OxfDmMY=), C6(v+SNVAiEe8m9X0W47l9tP8Zg4jUg2+RLqTSWdguLMFo=), D6(T9Wk1U9aj+X+NNH5mBBsHumTHTBfN85mvjy+HQKCFPs=)];0;v)
    Block B7(pCDmeRSHh8yGcTWPZGGqVDodKa9yoebkCskUZ8jKmZE=)(7008ms;[A6(luf/p0Q4l8TTqfmZdi5+ioJOR8d1BuOrPGXqxs1Vh4E=), B6(6LaW2zW6P4y9FOqsdh0Vx3vVrjFzbcu69iC1OxfDmMY=), C6(v+SNVAiEe8m9X0W47l9tP8Zg4jUg2+RLqTSWdguLMFo=), D6(T9Wk1U9aj+X+NNH5mBBsHumTHTBfN85mvjy+HQKCFPs=)];0;v)
    Block C7(dn90kGjh8rjZ/ekUeUXqLF2X8Uh+1/azMyCUpyZgcIA=)(7009ms;[A6(luf/p0Q4l8TTqfmZdi5+ioJOR8d1BuOrPGXqxs1Vh4E=), B6(6LaW2zW6P4y9FOqsdh0Vx3vVrjFzbcu69iC1OxfDmMY=), C6(v+SNVAiEe8m9X0W47l9tP8Zg4jUg2+RLqTSWdguLMFo=), D6(T9Wk1U9aj+X+NNH5mBBsHumTHTBfN85mvjy+HQKCFPs=)];0;v)
    Block D7(BaO6lh9kpHHjCqzxA1KYx/0fe7PCYxyOZxCEcbu6k/4=)(7010ms;[A6(luf/p0Q4l8TTqfmZdi5+ioJOR8d1BuOrPGXqxs1Vh4E=), B6(6LaW2zW6P4y9FOqsdh0Vx3vVrjFzbcu69iC1OxfDmMY=), C6(v+SNVAiEe8m9X0W47l9tP8Zg4jUg2+RLqTSWdguLMFo=), D6(T9Wk1U9aj+X+NNH5mBBsHumTHTBfN85mvjy+HQKCFPs=)];0;v)
}
```

### DagParser

Example Dag 
```
let dag_str = "DAG { 
        Round 0 : { 4 },
        Round 1 : { * },
        Round 2 : { * },
        Round 3 : { * },
        Round 4 : {
            A -> [-D3],
            B -> [*],
            C -> [*],
            D -> [*],
        },
        Round 5 : {
            A -> [*],
            B -> [*],
            C -> [A4],
            D -> [A4],
        },
        Round 6 : { * },
        Round 7 : { * },
        Round 8 : { * },
     }"
     
// will parse dag dsl & write interpreted dag to dag state.
test_dag_parser::parse_dag(dag_str)?
```

- `Round 0 : { n }`:  A special genesis case where `n` initializes how many authorities are to be include in the committee
- `{ * }`: Fully connected layer to ancestors from previous round
-  `A -> [-D3]`: Exclude specified ancestors for an authority
-  `A -> [D3]`: Include specified ancestor for an authority
-  `A -> [*]`: Include all specified ancestors for an authority
-  todo: add option for specifying equivocating slots



## Test Plan 

Still need to add more unit tests for `DagBuilder`

akichidis reviewed (commented):
This is really great work @arun-koshy  - looking forward to start using that tool!

Now, I've left some comments which have to do mostly with the usage patterns of the layer builder api because it seems that we are giving too much flexbility/responsibility to the user and under circumstances it could produce unexpected results. I don't believe that at this point is worth refining things too much, but at least guard against some usages.

Also, following our offsite discussion I start believing as well that we might be able to avoid using the DagState. Most of the access patterns seem to be simple. So that will allow us to decouple those two components. Flushing the produced blocks to DagState can be done via other ways.

I want to do another pass but those are my so far comments. Looking forward to hear your thoughts

akichidis commented on consensus/core/src/test_dag_builder.rs:
```diff
@@ -0,0 +1,449 @@
+                round = block.round();
+                dag_str.push_str(&format!("Round {round} : \n"));
+            }
+            dag_str.push_str(&format!("    Block {block:#?}\n"));
```
Hhm good point. Ok let's leave as is , not that important anyways

akichidis commented on consensus/core/src/test_dag_builder.rs:
```diff
@@ -0,0 +1,449 @@
+        self.leader_round = Some(leader_round);
+        self.fully_linked_ancestors = false;
+        self
+    }
```
Yeah that's another catch here, but I can't think a straightforward solution to that that doesn't require more refactoring to the API it self. Having comments for now on the methods might be good enough

akichidis commented on consensus/core/src/test_dag_builder.rs:
```diff
@@ -0,0 +1,449 @@
+        self.no_leader_block = true;
+        self.leader_offset = Some(leader_offset);
+        self
+    }
```
Latter option seems good to me, thanks for taking care of it!

akichidis commented:
> This is really great work @arun-koshy - looking forward to start using that tool!
> 
> Now, I've left some comments which have to do mostly with the usage patterns of the layer builder api because it seems that we are giving too much flexbility/responsibility to the user and under circumstances it could produce unexpected results. I don't believe that at this point is worth refining things too much, but at least guard against some usages.
> 
> Also, following our offsite discussion I start believing as well that we might be able to avoid using the DagState. Most of the access patterns seem to be simple. So that will allow us to decouple those two components. Flushing the produced blocks to DagState can be done via other ways.
> 
> I want to do another pass but those are my so far comments. Looking forward to hear your thoughts

@arun-koshy is this ready for re-review? I am fine with the changes I just want to confirm if using the `DagState` is the preferred way to go or you considered my earlier comment around it?

arun-koshy commented:
> > This is really great work @arun-koshy - looking forward to start using that tool!
> > Now, I've left some comments which have to do mostly with the usage patterns of the layer builder api because it seems that we are giving too much flexbility/responsibility to the user and under circumstances it could produce unexpected results. I don't believe that at this point is worth refining things too much, but at least guard against some usages.
> > Also, following our offsite discussion I start believing as well that we might be able to avoid using the DagState. Most of the access patterns seem to be simple. So that will allow us to decouple those two components. Flushing the produced blocks to DagState can be done via other ways.
> > I want to do another pass but those are my so far comments. Looking forward to hear your thoughts
> 
> @arun-koshy is this ready for re-review? I am fine with the changes I just want to confirm if using the `DagState` is the preferred way to go or you considered my earlier comment around it?

My fault, I read that and forgot to make the change on my last pass through. Removed the usage of DagState inside of DagBuilder so the control moves to the user of DagBuilder of when to persist. Added some methods to do this persist at the DagBuilder and LayerBuilder level. There could be some gotchas with the usage as there is no protection of calling this persist method multiple times and potentially writing duplicate blocks to DagState, but we can iterate on that part if needed.

mwtian commented on consensus/core/src/dag_state.rs:
```diff
@@ -681,18 +684,44 @@ impl DagState {
+    // roudn as a leader block.
+    // TODO: confirm pipelined & multi-leader cases work properly
+    #[cfg(test)]
+    pub(crate) fn get_all_uncommitted_leader_blocks(
```
Why not keep this function in test_dag.rs, or move it to test_dag_builder.rs or another location? It seems no DagState internal function is called.

I guess it is easier to implement this way, but it seems this function will not play well with a DagState used "normally". For example if the DagState flushes out some blocks, then iterating from round 1 will not get back any result.

mwtian commented on consensus/core/src/tests/universal_committer_tests.rs:
```diff
@@ -13,20 +13,27 @@ use crate::{
     let leader_round_wave_1 = committer.committers[0].leader_round(1);
     let voting_round_wave_2 = committer.committers[0].leader_round(2) + 1;
-    build_dag(context, dag_state, None, voting_round_wave_2);
+    dag_builder
```
Nice!

mwtian commented on consensus/core/src/test_dag_builder.rs:
```diff
@@ -0,0 +1,568 @@
+    // Configuration options applied to specified authorities
+    specified_authorities: Option<Vec<AuthorityIndex>>,
+    // Number of equivocating blocks per specified authority
+    equivocations: Option<usize>,
```
Maybe just default to 0?

mwtian commented on consensus/core/src/test_dag_builder.rs:
```diff
@@ -0,0 +1,568 @@
+    leader_round: Option<Round>,
+
+    // All ancestors will be linked to the current layer
+    fully_linked_ancestors: bool,
```
Maybe leave a TODO to turn them into enum.
